File: jupyter_ai_persona_manager/persona_events.py
"""Jupyter Events-based publishing of persona session information.

This replaces the previous Yjs-awareness mechanism (``persona_awareness.py``).
Instead of writing awareness slots, the persona manager and each persona emit
Jupyter Events:

- ``jupyter_ai_persona_manager/personas/v1`` -- the list of personas in a chat
  (published by the manager).
- ``jupyter_ai_persona_manager/persona_state/v1`` -- a single persona's session
  state: model configuration, general settings, usage, and slash commands
  (published by each persona on change).

Events are fire-and-forget, so a client that connects after an event was emitted
would miss it. Catch-up is handled by re-emitting the current state when a client
connects to the chat (see ``PersonaManager`` and the ``client_connected`` action
on jupyterlab_chat's ``room/v1`` event bus). The current values are kept in memory
here so they can be re-emitted at any time.

Each event carries the chat's stable id (``chat.get_id()``) so the frontend can
route it to the right chat via its chat model/context ``id``.
"""
from __future__ import annotations

from logging import Logger
from typing import TYPE_CHECKING, Any, Optional

from .awareness_models import (
    CommandOption,
    ModelConfiguration,
    PersonaOption,
    SettingConfiguration,
    Usage,
)

if TYPE_CHECKING:
    from jupyter_events import EventLogger


PERSONAS_EVENT_SCHEMA_ID = (
    "https://schema.jupyter.org/jupyter_ai_persona_manager/personas/v1"
)
PERSONA_STATE_EVENT_SCHEMA_ID = (
    "https://schema.jupyter.org/jupyter_ai_persona_manager/persona_state/v1"
)
PERSONA_SELECTED_EVENT_SCHEMA_ID = (
    "https://schema.jupyter.org/jupyter_ai_persona_manager/persona_selected/v1"
)

PERSONAS_EVENT_SCHEMA = {
    "$id": PERSONAS_EVENT_SCHEMA_ID,
    "version": "1",
    "title": "Persona list",
    "personal-data": True,
    "description": "The list of personas available in a chat.",
    "type": "object",
    "required": ["chat_id", "personas"],
    "properties": {
        "chat_id": {"type": "string", "description": "The chat's stable id (used by the frontend to scope events to a chat)."},
        "personas": {
            "type": "array",
            "items": {"type": "object"},
            "description": "Serialized PersonaOption objects.",
        },
    },
    "additionalProperties": False,
}

PERSONA_STATE_EVENT_SCHEMA = {
    "$id": PERSONA_STATE_EVENT_SCHEMA_ID,
    "version": "1",
    "title": "Persona session state",
    "personal-data": True,
    "description": (
        "A single persona's model, settings, usage, slash commands, and "
        "whether it is currently processing a message."
    ),
    "type": "object",
    "required": ["chat_id", "persona_id"],
    "properties": {
        "chat_id": {"type": "string", "description": "The chat's stable id (used by the frontend to scope events to a chat)."},
        "persona_id": {"type": "string", "description": "The persona's stable id."},
        "model": {"type": "object"},
        "settings": {"type": "array", "items": {"type": "object"}},
        "usage": {"type": "object"},
        "slash_commands": {"type": "array", "items": {"type": "object"}},
        "processing": {
            "type": "boolean",
            "description": (
                "Whether the persona is currently processing a message. The "
                "frontend uses this to enable the stop button."
            ),
        },
    },
    "additionalProperties": False,
}

PERSONA_SELECTED_EVENT_SCHEMA = {
    "$id": PERSONA_SELECTED_EVENT_SCHEMA_ID,
    "version": "1",
    "title": "Persona selected",
    "personal-data": True,
    "description": (
        "Emitted by a client when a persona is selected in a chat, so the "
        "server can prepare it eagerly."
    ),
    "type": "object",
    "required": ["chat_id", "persona_id"],
    "properties": {
        "chat_id": {"type": "string", "description": "The chat's stable id the selection happened in."},
        "persona_id": {"type": "string", "description": "The stable id of the selected persona."},
    },
    "additionalProperties": False,
}


def register_persona_event_schemas(event_logger: "EventLogger") -> None:
    """Register the persona event schemas on ``event_logger`` (idempotent)."""
    for schema in (
        PERSONAS_EVENT_SCHEMA,
        PERSONA_STATE_EVENT_SCHEMA,
        PERSONA_SELECTED_EVENT_SCHEMA,
    ):
        try:
            event_logger.register_event_schema(schema)
        except Exception:  # pragma: no cover - already registered / defensive
            pass


class PersonaManagerSessionState:
    """The persona-manager's session state that is published to clients via the
    events API: the chat's persona list.

    Replaces ``PersonaManagerAwareness``. Holds the last-published list in memory
    so it can be re-emitted for catch-up when a new client connects.
    """

    def __init__(
        self,
        *,
        event_logger: Optional["EventLogger"],
        chat_id: str,
        log: Logger,
    ):
        self._event_logger = event_logger
        self._chat_id = chat_id
        self._log = log
        self._personas: list[PersonaOption] = []

    @property
    def personas(self) -> list[PersonaOption]:
        return list(self._personas)

    @personas.setter
    def personas(self, personas: list[PersonaOption]) -> None:
        self._personas = list(personas)
        self.publish()

    def publish(self) -> None:
        """(Re-)emit the current persona list. Used both on change and for
        catch-up when a client connects."""
        if self._event_logger is None:
            return
        try:
            self._event_logger.emit(
                schema_id=PERSONAS_EVENT_SCHEMA_ID,
                data={
                    "chat_id": self._chat_id,
                    "personas": [p.model_dump() for p in self._personas],
                },
            )
        except Exception:  # pragma: no cover - defensive
            self._log.exception("Failed to emit persona list event")


class PersonaSessionState:
    """A persona's session state that is published to clients via the events API:
    its model configuration, general settings, usage, and slash commands.

    Replaces ``PersonaAwareness``. The typed properties keep the last value in
    memory and emit a ``persona_state`` event carrying only the changed
    attribute, so consumers merge live updates onto their existing state;
    :meth:`publish` emits the full state for catch-up.
    """

    def __init__(
        self,
        *,
        event_logger: Optional["EventLogger"],
        persona_id: str,
        chat_id: str,
        log: Logger,
    ):
        self._event_logger = event_logger
        self._chat_id = chat_id
        self._persona_id = persona_id
        self._log = log
        self._model = ModelConfiguration()
        self._settings: list[SettingConfiguration] = []
        self._usage = Usage()
        self._slash_commands: list[CommandOption] = []
        self._processing = False

    @property
    def id(self) -> str:
        return self._persona_id

    @property
    def model(self) -> ModelConfiguration:
        return self._model

    @model.setter
    def model(self, model: ModelConfiguration) -> None:
        self._model = model
        self._emit({"model": model.model_dump()})

    @property
    def settings(self) -> list[SettingConfiguration]:
        return self._settings

    @settings.setter
    def settings(self, settings: list[SettingConfiguration]) -> None:
        self._settings = settings
        self._emit({"settings": [s.model_dump() for s in settings]})

    @property
    def usage(self) -> Usage:
        return self._usage

    @usage.setter
    def usage(self, usage: Usage) -> None:
        self._usage = usage
        self._emit({"usage": usage.model_dump()})

    @property
    def slash_commands(self) -> list[CommandOption]:
        return self._slash_commands

    @slash_commands.setter
    def slash_commands(self, commands: list[CommandOption]) -> None:
        self._slash_commands = commands
        self._emit({"slash_commands": [c.model_dump() for c in commands]})

    @property
    def processing(self) -> bool:
        return self._processing

    @processing.setter
    def processing(self, processing: bool) -> None:
        # Only publish on an actual transition, so repeated no-op assignments
        # (e.g. from serialized message processing) don't emit redundant events.
        if self._processing == processing:
            return
        self._processing = processing
        self._emit({"processing": processing})

    def to_data(self) -> dict[str, Any]:
        """The persona's full current state, including the routing ids. Used by
        :meth:`publish` for the catch-up snapshot."""
        return {
            "chat_id": self._chat_id,
            "persona_id": self._persona_id,
            "model": self._model.model_dump(),
            "settings": [s.model_dump() for s in self._settings],
            "usage": self._usage.model_dump(),
            "slash_commands": [c.model_dump() for c in self._slash_commands],
            "processing": self._processing,
        }

    def _emit(self, fields: dict[str, Any]) -> None:
        """Emit a ``persona_state`` event carrying ``fields`` plus the routing
        ids (``chat_id``/``persona_id``).

        Field setters emit only the attribute that changed; the frontend merges
        each event into the persona's existing state, treating absent keys as
        unchanged. This avoids re-sending the whole snapshot (model, settings,
        slash commands) on every usage tick. :meth:`publish` sends the full
        snapshot for catch-up when a client connects.
        """
        if self._event_logger is None:
            return
        try:
            self._event_logger.emit(
                schema_id=PERSONA_STATE_EVENT_SCHEMA_ID,
                data={
                    "chat_id": self._chat_id,
                    "persona_id": self._persona_id,
                    **fields,
                },
            )
        except Exception:  # pragma: no cover - defensive
            self._log.exception("Failed to emit persona state event")

    def publish(self) -> None:
        """Emit the persona's full current state. Used for catch-up when a
        client connects; individual field changes emit only the changed field
        (see :meth:`_emit`)."""
        data = self.to_data()
        # `_emit` re-adds the routing ids, so drop them from the field set here.
        data.pop("chat_id", None)
        data.pop("persona_id", None)
        self._emit(data)

    def shutdown(self) -> None:
        """Symmetry with the old awareness slot's shutdown. Events are
        fire-and-forget, so there is no retained slot to clear; this is a no-op
        kept so callers need not special-case it."""
        return None

File: src/persona-events.ts
/**
 * Frontend persona session state, delivered over Jupyter Events (via the
 * `ServiceManager` event bus), replacing the previous Yjs-awareness channel.
 *
 * Flow: the server emits `personas` / `persona_state` events (each carrying the
 * chat's stable `chat_id`); the `PersonaSessionRegistry` routes them to the
 * per-chat `PersonaManagerSessionState`, which holds the persona list and a
 * `PersonaSessionState` per persona and fires a Lumino `changed` signal. React
 * components listen to that signal. When a chat closes, its state is discarded.
 *
 * Names mirror the Python package (`PersonaManagerSessionState`,
 * `PersonaSessionState`).
 */
import { Event } from '@jupyterlab/services';
import { Token } from '@lumino/coreutils';
import { IDisposable } from '@lumino/disposable';
import { ISignal, Signal } from '@lumino/signaling';

import {
  CommandOption,
  EMPTY_MODEL_CONFIGURATION,
  EMPTY_USAGE,
  ModelConfiguration,
  PersonaOption,
  SettingConfiguration,
  Usage
} from './awareness';

export const PERSONAS_EVENT_SCHEMA_ID =
  'https://schema.jupyter.org/jupyter_ai_persona_manager/personas/v1';
export const PERSONA_STATE_EVENT_SCHEMA_ID =
  'https://schema.jupyter.org/jupyter_ai_persona_manager/persona_state/v1';
export const PERSONA_SELECTED_EVENT_SCHEMA_ID =
  'https://schema.jupyter.org/jupyter_ai_persona_manager/persona_selected/v1';

/** The wire shape of a `persona_state` event. */
export type PersonaStatePayload = {
  chat_id?: string;
  persona_id?: string;
  model?: ModelConfiguration;
  settings?: SettingConfiguration[];
  usage?: Usage;
  slash_commands?: CommandOption[];
  processing?: boolean;
};

/** The wire shape of a `personas` event. */
type PersonasPayload = {
  chat_id?: string;
  personas?: PersonaOption[];
};

/**
 * One persona's live session state, built from a `persona_state` event.
 * Immutable: each update produces a new instance, so React consumers see a new
 * reference and re-render.
 */
export class PersonaSessionState {
  constructor(
    public readonly id: string,
    payload: PersonaStatePayload = {}
  ) {
    this.model = payload.model ?? EMPTY_MODEL_CONFIGURATION;
    this.settings = payload.settings ?? [];
    this.usage = { ...EMPTY_USAGE, ...(payload.usage ?? {}) };
    this.slash_commands = payload.slash_commands ?? [];
    this.processing = payload.processing ?? false;
  }

  readonly model: ModelConfiguration;
  readonly settings: SettingConfiguration[];
  readonly usage: Usage;
  readonly slash_commands: CommandOption[];
  /** Whether the persona is currently processing a message. */
  readonly processing: boolean;

  /**
   * Merge a `persona_state` event payload onto this state, returning a new
   * instance. Attributes the event omits are carried forward unchanged, so a
   * partial event (e.g. usage-only) replaces only what it carries. A new
   * reference is returned so React consumers re-render.
   */
  withUpdate(payload: PersonaStatePayload): PersonaSessionState {
    return new PersonaSessionState(this.id, {
      model: payload.model ?? this.model,
      settings: payload.settings ?? this.settings,
      usage: payload.usage ?? this.usage,
      slash_commands: payload.slash_commands ?? this.slash_commands,
      processing: payload.processing ?? this.processing
    });
  }
}

/**
 * The per-chat manager session state: the persona list plus a
 * `PersonaSessionState` per persona. Fires `changed` whenever the persona list
 * or any persona's state updates, so React components re-render.
 */
export class PersonaManagerSessionState implements IDisposable {
  constructor(public readonly chatId: string) {}

  /** Emits whenever the persona list or a persona's state changes. */
  get changed(): ISignal<this, void> {
    return this._changed;
  }

  /** Whether one or several personas have been registered. */
  get ready(): boolean {
    return this._personasReceived;
  }

  /** The personas advertised in this chat (backend + frontend). */
  get personas(): PersonaOption[] {
    const personas = [
      ...this._backendPersonas,
      ...this._frontendPersonas.values()
    ];
    personas.sort((a, b) => (a.name >= b.name ? 1 : -1));
    return personas;
  }

  /** A persona's session state, or undefined if it has not published yet. */
  getPersona(id: string): PersonaSessionState | undefined {
    return this._states.get(id);
  }

  /**
   * Whether any persona in this chat is currently processing a message. The
   * stop button uses this to enable itself, since a persona can be processing
   * (thinking, running tools, awaiting an agent turn) without actively writing.
   */
  get processing(): boolean {
    for (const state of this._states.values()) {
      if (state.processing) {
        return true;
      }
    }
    return false;
  }

  /** Apply a `personas` event payload, keeping any registered frontend personas. */
  updateBackendPersonas(personas: PersonaOption[]): void {
    this._personasReceived = true;
    this._backendPersonas = personas;
    this._changed.emit();
  }

  /**
   * Register a frontend-only persona (one with no backend counterpart). It
   * survives subsequent `updatePersonas` calls and immediately marks the list
   * as ready, so the toolbar never shows the loading placeholder when only
   * frontend personas are available.
   */
  registerFrontendPersona(persona: PersonaOption): void {
    this._personasReceived = true;
    this._frontendPersonas.set(persona.id, persona);
    this._changed.emit();
  }

  /**
   * Unregister a frontend persona from its ID.
   */
  unregisterFrontendPersona(personaId: string): void {
    this._frontendPersonas.delete(personaId);
    this._changed.emit();
  }

  /**
   * Apply a `persona_state` event payload for one persona, merging it onto the
   * persona's existing state (absent attributes are left unchanged).
   */
  updatePersonaState(personaId: string, payload: PersonaStatePayload): void {
    const prev = this._states.get(personaId);
    const next = prev
      ? prev.withUpdate(payload)
      : new PersonaSessionState(personaId, payload);
    this._states.set(personaId, next);
    this._changed.emit();
  }

  get isDisposed(): boolean {
    return this._isDisposed;
  }

  dispose(): void {
    if (this._isDisposed) {
      return;
    }
    this._isDisposed = true;
    this._states.clear();
    this._frontendPersonas.clear();
    Signal.clearData(this);
  }

  private _personasReceived = false;
  private _backendPersonas: PersonaOption[] = [];
  private _frontendPersonas = new Map<string, PersonaOption>();
  private _states = new Map<string, PersonaSessionState>();
  private _isDisposed = false;
  private _changed = new Signal<this, void>(this);
}

/**
 * Routes persona events to the correct per-chat `PersonaManagerSessionState`,
 * creating one on demand and discarding it when the chat closes.
 *
 * A single instance is created by the plugin and shared with the toolbar
 * controls and the slash-command provider.
 */
export class PersonaSessionRegistry {
  constructor(events: Event.IManager) {
    // The ServiceManager event bus (JupyterLab >= 4.0) exposes a single shared
    // stream of all Jupyter Events; filter it by schema id to route the two
    // persona event types. This supersedes the former `jupyterlab-eventlistener`
    // dependency.
    events.stream.connect((_, emission) => {
      if (emission.schema_id === PERSONAS_EVENT_SCHEMA_ID) {
        void this._onPersonasEvent(emission);
      } else if (emission.schema_id === PERSONA_STATE_EVENT_SCHEMA_ID) {
        void this._onPersonaStateEvent(emission);
      }
    });
  }

  /**
   * Get (or create) the manager session state for a chat. Components call this
   * with their chat's stable id (`IChatModel.id` / `IChatContext.id`).
   */
  get(chatId: string): PersonaManagerSessionState {
    let state = this._byChatId.get(chatId);
    if (!state) {
      state = new PersonaManagerSessionState(chatId);
      this._byChatId.set(chatId, state);
    }
    return state;
  }

  /** Whether a manager session state exists for `chatId` (without creating one). */
  has(chatId: string): boolean {
    return this._byChatId.has(chatId);
  }

  /**
   * Register a frontend-only persona for a chat. Shorthand for
   * `registry.get(chatId).registerFrontendPersona(persona)`.
   */
  registerFrontendPersona(chatId: string, persona: PersonaOption): void {
    this.get(chatId).registerFrontendPersona(persona);
  }

  /**
   * Update a persona's state for a chat. Shorthand for
   * `registry.get(chatId).updatePersonaState(personaId, payload)`.
   */
  updatePersonaState(
    chatId: string,
    personaId: string,
    payload: PersonaStatePayload
  ): void {
    this.get(chatId).updatePersonaState(personaId, payload);
  }

  /**
   * Discard a chat's session state and free its memory. Called when the client
   * closes the chat (wired to the chat model's `disposed` signal).
   */
  discard(chatId: string): void {
    const state = this._byChatId.get(chatId);
    if (state) {
      state.dispose();
      this._byChatId.delete(chatId);
    }
  }

  private _onPersonasEvent = async (event: Event.Emission): Promise<void> => {
    const data = event as PersonasPayload;
    if (!data.chat_id) {
      return;
    }
    this.get(data.chat_id).updateBackendPersonas(
      Array.isArray(data.personas) ? data.personas : []
    );
  };

  private _onPersonaStateEvent = async (
    event: Event.Emission
  ): Promise<void> => {
    const data = event as PersonaStatePayload;
    if (!data.chat_id || !data.persona_id) {
      return;
    }
    this.get(data.chat_id).updatePersonaState(data.persona_id, data);
  };

  private _byChatId = new Map<string, PersonaManagerSessionState>();
}

/**
 * Plugin token for the shared `PersonaSessionRegistry`. Consumed by the toolbar
 * controls and the slash-command provider.
 */
export const IPersonaSessionRegistry = new Token<PersonaSessionRegistry>(
  '@jupyter-ai/persona-manager:IPersonaSessionRegistry'
);
